---
title: "Class 19: Investigating Pertussis Resurgence"
author: "Olivia Chu"
format: html
---

# Investigating pertussis cases by year

The CDC tracks cases of Pertussis in the US. We can get their data via web-scrapping.

> Q1. With the help of the R “addin” package datapasta, assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r echo=FALSE}
cdc <- data.frame(
  Year = c(
    1922L,
    1923L,
    1924L,
    1925L,
    1926L,
    1927L,
    1928L,
    1929L,
    1930L,
    1931L,
    1932L,
    1933L,
    1934L,
    1935L,
    1936L,
    1937L,
    1938L,
    1939L,
    1940L,
    1941L,
    1942L,
    1943L,
    1944L,
    1945L,
    1946L,
    1947L,
    1948L,
    1949L,
    1950L,
    1951L,
    1952L,
    1953L,
    1954L,
    1955L,
    1956L,
    1957L,
    1958L,
    1959L,
    1960L,
    1961L,
    1962L,
    1963L,
    1964L,
    1965L,
    1966L,
    1967L,
    1968L,
    1969L,
    1970L,
    1971L,
    1972L,
    1973L,
    1974L,
    1975L,
    1976L,
    1977L,
    1978L,
    1979L,
    1980L,
    1981L,
    1982L,
    1983L,
    1984L,
    1985L,
    1986L,
    1987L,
    1988L,
    1989L,
    1990L,
    1991L,
    1992L,
    1993L,
    1994L,
    1995L,
    1996L,
    1997L,
    1998L,
    1999L,
    2000L,
    2001L,
    2002L,
    2003L,
    2004L,
    2005L,
    2006L,
    2007L,
    2008L,
    2009L,
    2010L,
    2011L,
    2012L,
    2013L,
    2014L,
    2015L,
    2016L,
    2017L,
    2018L,
    2019L
  ),
  Cases = c(
    107473,
    164191,
    165418,
    152003,
    202210,
    181411,
    161799,
    197371,
    166914,
    172559,
    215343,
    179135,
    265269,
    180518,
    147237,
    214652,
    227319,
    103188,
    183866,
    222202,
    191383,
    191890,
    109873,
    133792,
    109860,
    156517,
    74715,
    69479,
    120718,
    68687,
    45030,
    37129,
    60886,
    62786,
    31732,
    28295,
    32148,
    40005,
    14809,
    11468,
    17749,
    17135,
    13005,
    6799,
    7717,
    9718,
    4810,
    3285,
    4249,
    3036,
    3287,
    1759,
    2402,
    1738,
    1010,
    2177,
    2063,
    1623,
    1730,
    1248,
    1895,
    2463,
    2276,
    3589,
    4195,
    2823,
    3450,
    4157,
    4570,
    2719,
    4083,
    6586,
    4617,
    5137,
    7796,
    6564,
    7405,
    7298,
    7867,
    7580,
    9771,
    11647,
    25827,
    25616,
    15632,
    10454,
    13278,
    16858,
    27550,
    18719,
    48277,
    28639,
    32971,
    20762,
    17972,
    18975,
    15609,
    18617
  )
)
```

```{r}
library(ggplot2)

baseplot <- ggplot(cdc) +
  aes(Year, Cases) + 
  geom_point() +
  geom_line() + 
  labs(title="Cases of Pertussis in US from 1920 to 2019",
      subtitle="Data from CDC")

baseplot
```

# A tale of two vaccines (wP a& aP)

> Q2. Using the ggplot geom_vline() function, add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
baseplot + 
  geom_vline(xintercept=1946, col="cornflowerblue", linetype=2) +
  geom_text(aes(1950, 2.8e+05), label="wP", col="cornflowerblue", size=3) +
  geom_vline(xintercept=1996, col="hotpink", linetype=2) +
  geom_text(aes(2000, 2.8e+05), label="aP", col="hotpink", size=3)
```

When the wP vaccine was introduced in 1946, there was a significant drop in pertussis cases. The number of cases remained low up until the introduction of the aP vaccine in 1996. After the switch to the aP vaccine, we can observe a slight increase in pertussis cases.

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

After the switch to the aP vaccine, we can see that although case numbers remained low, there was a slight increase in pertussis cases. Possible causes for this increase include: sensitive PCR testing, vaccine hesitancy (people may not have wanted the aP vaccine after the success of the wP vaccine), bacterial evolution (the aP vaccine was less effective against pertussis), and less immunity within the community.

**Key-Point:** Despite high levels of acellular pertussis (aP) vaccination, the United States and other countries are now experiencing a significant resurgence in pertussis cases with large outbreaks now once again a major public health concern.

# The CMI-PB project

The CMI-PB project is collecting data on aP and wP individuals and their immune response to infection and/or booster shots.

CMI-PB returns data from its API in JSON format (like most APIs). We will use the jsonlite package to get data from this API.

```{r}
library(jsonlite)
```

```{r}
subject <- read_json("http://cmi-pb.org/api/subject", 
                     simplifyVector = TRUE)

head(subject)
```

> Q4. How may aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

There are 47 aP and 49 wP infancy vaccinated subjects in the dataset.

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```

There are 30 Male and 66 Female subjects/patients in this dataset.

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$race, subject$biological_sex)
```

See table above for race and biological sex breakdown. 

# Side-Note: Working with dates

```{r}
library(lubridate)
```

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

First, calculate the age in years of all subjects:

```{r}
age_days <- today() - ymd(subject$year_of_birth)
age_years <- time_length(age_days, "years")
subject$age <- age_years
```

Next, find the average age of all individuals:

```{r}
mean(subject$age)
```

Calculate the average age of wP and aP individuals:

```{r}
library(dplyr)

wp_age <- filter(subject, infancy_vac == "wP")$age
ap_age <- filter(subject, infancy_vac == "aP")$age

mean(wp_age)
mean(ap_age)
```

The average age of wP individuals is 36.36 years, and the average age of aP individuals is 26.51 years.

```{r}
# T-test
t.test(wp_age, ap_age)
```

This data is significant because my calculated p-value from my t-test is smaller than 0.05.

> Q8. Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)

age_at_boost <- time_length(int, "year")

age_at_boost
```
The value "age_at_boost" lists the age of all the individuals at the time of their boost.

> Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```

Yes, based on the facted boxplot, I do think these two groups are significantly different due to the vastly different counts distributions. the faceted boxplot for aP is skewed towards the left whereas the faceted boxplot for wP is more spread out and in the middle and towards the right.

# Joining multiple tables

Read the specimen and ab_titer tables into R and store the data as specimen and titer named data frames.

```{r}
# Complete the API URLs...
specimen <- read_json("http://cmi-pb.org/api/specimen", simplifyVector = TRUE) 

titer <- read_json("http://cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
```

```{r}
head(specimen)
```

```{r}
head(titer)
```

To know whether a given specimen_id comes from an aP or wP individual, we need to link (a.k.a. “join” or merge) our specimen and subject data frames. The excellent dplyr package (that we have used previously) has a family of `join()` functions that can help us with this common task:

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
meta <- inner_join(specimen, subject)

dim(meta)
head(meta)
```

> Q10. Now using the same procedure, join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)

dim(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```
We have 6698 IgE, 1413 IgG, and 6141 of IgG1, IgG2, IgG3, and IgG4 each. 

> Q12. What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit)
```

The number of visit 8 specimens is very low compared to other visits. The number of specimens had a sharp decreased from the 3000's to 80 during visit 8. This low number could be because data is still being collected and this project is still ongoing.

# Examine IgG1 Ab titer levels

Now, use our joined/merged/linked abdata dataset `filter()` for IgG1 isotype and exclude the small number of visit 8 entries.

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

The antigens that show differences in the level of IgG1 antibody titers recognizing them over time are PT, FIM2/3, and FHA. These antigens show differences (and not others) because they are all in the aP vaccine.

Additional antigens that show differences include TT, PRN, and DT.

We can attempt to examine differences between wP and aP here by setting color and/or facet values of the plot to include infancy_vac status

```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw()
```

Another version of this plot adding infancy_vac to the faceting:

```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

> and the same for antigen=="FIM2/3"

```{r}
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

> Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?

Compared to Measles, the FIM2/3 antigen sees a much more rapid increase in IgG1 antibody titers over time (this can also be observed by the larger values on the y-axis of FIM2/3). The Measles antigen saw a very small (maybe even neglible) change in antibody titers over this same time course.

> Q17. Do you see any clear difference in aP vs. wP responses?

There is a clear difference in aP (red) and wP (teal) responses (i.e. antigen levels).

When looking at Measles, wp response was greater than that of aP response. Over time, aP response increased and was able to catch up with wP response. 

When looking at the FIM2/3 plot, we can see that the wP response was much higher than aP response, with a steady increase in antigen levels from visit 1 to 3. However, from visit 4 to visit 7, aP response saw a much greater/rapid increase in antigen levels and it eventually overtakes wP response by the end of visit 7.

# Obtaining CMI-PB RNASeq data

We will read available RNA-Seq data for this gene into R and investigate the time course of its gene expression values.

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE)
```

To facilitate further analysis, we need to “join” the rna expression data with our metadata meta, which is itself a join of sample and specimen data. This will allow us to look at this genes TPM expression values over aP/wP status and at different visits (i.e. times):

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```
> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q19.: What do you notice about the expression of this gene (i.e. when is it at its maximum level)?

The expression of this gene reaches its maximum level on visit 4.

> Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

No, this pattern in time does not match the trend of antibody titer data. In the antibody titer data, there is a steady increase of antigen levels, which peaked at visit 5 (FIM2/3). Expression of this gene peaked at visit 4. This is because when cells make antibodies, antibodies are able to live longer in the body, thus are still detectable after longer periods of time. 

We can dig deeper and color and/or facet by infancy_vac status:

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

There is no obvious wP vs. aP differences here even if we focus in on a particular visit:

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```